File: lemur-libs/lemur-contract/src/main/java/io/github/malczuuu/lemur/contract/rest/CursorResult.java
package io.github.malczuuu.lemur.contract.rest;

import io.github.malczuuu.lemur.contract.TransportResponse;
import java.util.List;
import org.jspecify.annotations.Nullable;

public record CursorResult<T>(List<T> content, @Nullable Links links, @Nullable Long totalElements)
    implements TransportResponse {

  public record Links(@Nullable String next) implements TransportResponse {}
}

File: lemur-libs/lemur-contract/src/main/kotlin/io/github/malczuuu/lemur/contract/rest/CursorResult.kt
package io.github.malczuuu.lemur.contract.rest

import io.github.malczuuu.lemur.contract.TransportResponse
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.PositiveOrZero
import jakarta.validation.constraints.Size

/**
 * See [TransportResponse] for explanation of why fields are nullable.
 */
data class CursorResult<T>(

    @field:NotNull
    val content: List<T>?,

    @field:Valid
    val links: Links?,

    @field:PositiveOrZero
    val totalElements: Long?,
) : TransportResponse {
    data class Links(@field:Size(min = 1) val next: String? = null) : TransportResponse
}
